File: Assign_1/src/FixedPointNumber.java
// author of program:  Ramish
// Description:  This program is a fixed point number that can be added, subtracted, multiplied, divided, and converted to a different q value
public class FixedPointNumber {
    final private int intVal;
    final private int qVal ;

    // no method in class can call this constructor
    private FixedPointNumber(int intVal, int qVal) {
        this.intVal = (int) (intVal * Math.pow(2, qVal));;
        this.qVal = qVal;
    }

    // use as default for constructor
    public FixedPointNumber(double x, int q) {
        this.intVal = (int) (x * Math.pow(2, q));
        this.qVal = q;
    }

    // to be called to string method
    public double toDouble() {
        return  (this.intVal / Math.pow(2, this.qVal));
    }

    //convert from one fixed point to another
    public FixedPointNumber toQval(int q) {
        FixedPointNumber number = null;
        if (this.intVal > q) {
            number = new FixedPointNumber(intVal >> (this.qVal - q), q);
        } else if (qVal < q) {
            number = new FixedPointNumber(intVal << (q - qVal), q);
        }
        return number;
    }

    // fixed point to string
    public String toString() {
        return String.format("%d,%d: %.6f", this.intVal, this.qVal, this.toDouble());
    }

    // two fixed point numbers equal
    public boolean equals(FixedPointNumber x) {
        return (this.intVal == x.intVal && this.qVal == x.qVal);
    }

    // add two fixed point numbers
    public FixedPointNumber plus(FixedPointNumber x, int q) {
        return new FixedPointNumber(this.toDouble() + x.toDouble(), q);
    }
}

File: Assign_1_Ruby/lib/FixedPointNumber.rb
class FixedPointNumber
  @int_val = 0
  @q_val = 12
  attr_reader :int_val, :q_val
  def initialize(int_val, q_val)
    #Check if intVal is an integer
    if Integer === int_val then
      @int_val = int_val
      @q_val = q_val
      # Check if intVal is a float
    elsif Float === int_val then
        @int_val = (int_val * (2 ** q_val)).to_i
        @q_val = q_val
    end
  end

  # convert to double
  def to_double
    return @int_val.to_f / (2 ** @q_val)
  end

  #convert from one fixed point to another
  def to_Q_val(q_val)
    if (@int_val > q_val)
      return FixedPointNumber.new(@int_val >> (@q_val - q_val), q_val)
    elsif (@q_val < q_val) then
      return FixedPointNumber.new(@int_val << (q_val - @q_val), q_val)
    end
  end
  #
  # fixed point to string
  def to_s
    return @int_val.to_s + ", " + @q_val.to_s + ": " + + sprintf(" %.6f", self.to_double)
  end

  # #two fixed point numbers are equal if they have the same integer value and the same q value
  def equals(other)
    if (other.is_a?(FixedPointNumber)) then
      return @int_val == other.int_val && @q_val == other.q_val
    end
  end

  #add two fixed point numbers
  def add(other,q)
    return FixedPointNumber.new(@int_val + other.int_val, q)
  end

end

# a = FixedPointNumber.new(2.5, 12)
# b = FixedPointNumber.new(3.5, 12)
# puts a.to_s
# puts b.to_s
# puts a.equals(b)
# puts a.add(b, 12).to_s
# puts(FixedPointNumber.new(1.5, 2).to_s)